File: pilha/Pilha Dinâmica/Java/Pilha Dinâmica/PilhaDinamica.java
public class PilhaDinamica implements Empilhavel {
	private int tamanho;
	private int quantidade;
	private NoDuplo ponteiroTopo;
	
	public PilhaDinamica(int tamanho) {
		this.tamanho = tamanho;
		quantidade = 0;
		ponteiroTopo = null;
	}

	public PilhaDinamica() {
		this(10);
	}

	@Override
	public void empilhar(Object dado) {
		if (!estaCheia()) {
			NoDuplo noTemporario = new NoDuplo();
			noTemporario.setDado(dado);
			noTemporario.setAnterior(ponteiroTopo);
			if (!estaVazia()) {
				ponteiroTopo.setProximo(noTemporario);
			}
			ponteiroTopo = noTemporario;
			quantidade++;
		} else {
			System.err.println("Pilha Cheia!");
		}
	}
	
	@Override
	public Object desempilhar() {
		Object dadoTopo = null;
		if (!estaVazia()) {
			dadoTopo = ponteiroTopo.getDado();
			ponteiroTopo = ponteiroTopo.getAnterior();
			quantidade--;
			if (!estaVazia()) {
				ponteiroTopo.setProximo(null);
			}
		} else {
			System.err.println("Pilha Vazia!");
		}
		return dadoTopo;
	}
	
	@Override
	public Object topo() {
		Object dadoTopo = null;
		if (!estaVazia()) {
			dadoTopo = ponteiroTopo.getDado();
		} else {
			System.err.println("Pilha Vazia!");
		}
		return dadoTopo;
	}
	
	@Override
	public boolean estaVazia() {
		return (quantidade == 0);
	}
	
	@Override
	public boolean estaCheia() {
		return (quantidade == tamanho);
	}
	
	@Override
	public String imprimir() {
		NoDuplo ponteiroAuxiliar = ponteiroTopo;
		String resultado = "[";
		for (int i = quantidade - 1; i >= 0; i--) {
			if (i == 0) {
				resultado += ponteiroAuxiliar.getDado();
			} else {
				resultado += ponteiroAuxiliar.getDado() + ",";
			}
			ponteiroAuxiliar = ponteiroAuxiliar.getAnterior();
		}
		return resultado + "]";
	}
}

File: pilha/dinamica/Kotlin/Pilha/PilhaDinamica.kt
class PilhaDinamica(private val tamanho: Int = 10) : Empilhavel {

	private var ponteiroTopo: NoDuplo? = null
	private var quantidade = 0
 
	override fun empilhar(dado: Any?) {
		if (!estaCheia()) {
			var noTemp = NoDuplo(dado)
			//noTemp.dado = dado
			noTemp.anterior = ponteiroTopo
			if (!estaVazia())
				ponteiroTopo?.proximo = noTemp

			ponteiroTopo = noTemp
			quantidade = quantidade.inc()
		} else {
			println("Pilha Cheia!")
		}
	}
	
	override fun desempilhar(): Any? {
		var dadoTopo: Any? = null
		if (!estaVazia()) {
			dadoTopo = ponteiroTopo?.dado
			ponteiroTopo = ponteiroTopo?.anterior
			quantidade = quantidade.dec()
			if (!estaVazia())
				ponteiroTopo?.proximo = null
				
		} else {
			println("Pilha Vazia!")
		}
		
		return dadoTopo
	}
	
	override fun espiar(): Any? {
		var dadoTopo: Any? = null
		if (!estaVazia())
			dadoTopo = ponteiroTopo?.dado
		else
			println("Pilha Vazia!")

		return dadoTopo
	}
	
	override fun atualizar(dado: Any?) {
		if (!estaVazia())
			ponteiroTopo?.dado = dado
		else
			println("Pilha Vazia!")
	}

	override fun estaCheia(): Boolean {
		return quantidade == tamanho
	}
	
	override fun estaVazia(): Boolean {
		return quantidade == 0
	}
	
	override fun imprimir(): String {
		var ponteiroAuxiliar = ponteiroTopo
		var resultado = "["
		for (i in 0 until quantidade) {
			resultado += "${ponteiroAuxiliar?.dado}"
			if (i != quantidade-1)
				resultado += ","
			
			ponteiroAuxiliar = ponteiroAuxiliar?.anterior
		}
		return "$resultado]"
	}
}